Implements AddTraits and RemoveTraits procs for adding/removing multiple traits + swag unit test

san7890 opened the pull request:
## About The Pull Request

On the tin, doing it like this means we can reduce our overall line fingerprint whenever we have to add two or more traits from the same source on the same target. Especially helps when we get to the 4+ range of traits, a breath of fresh air even.

Doesn't mean we have to do for loops, as that's already handled within the define as well. I replaced some of the checks with `length()` checks, let me know if I should switch it over to something else (maybe `islist()`)? We stack_trace whenever we're not passed a list reference on purpose, and sometimes var/lists are null by default (or just empty, making this redundant).
## Why It's Good For The Game

I commonly feel the urge to write "use `AddTraits()`" or something in reviews, then am sad when I remember it doesn't exist. I will no longer be sad.

Can ensure a lot more trait safety as well by using static lists- when both ADD_TRAIT_LIST and REMOVE_TRAIT_LIST re-use the same list, you are confident (from a static point of view) that everything that you want to be adding/removing works.

I may have missed a few things where this could be used, but both macros implemented in this PR still use the same framework that was being used in the last four years- so stuff won't break if left untouched. Just a nifty new tool for developers.

also fixed up some code in the area, numerous bugs were found and exploded
## Changelog
Nothing that really concerns players (aside from potential breaks, but I'll audit those myself in the coming 24 hours).

i also didn't use any regexes, this was manual work. soz

bonus unit test: mainly because i wanted to make sure mixing/matching ADD_TRAIT/REMOVE_TRAIT with ADD_TRAITS/REMOVE_TRAITS would still work, but i decided I should test all of the "basic" add/remove trait stuff we have juuust in case

san7890 committed:
adds the defines

san7890 committed:
comment update

san7890 committed:
first pass

san7890 committed:
more pass

san7890 committed:
anotha pass

san7890 committed:
last of the ADD_TRAIT

san7890 committed:
fixes up all the forloops

length() checking to be safe in some circumstances

san7890 committed:
oopsy doopsy

san7890 committed:
static listifies hulk traits

san7890 committed:
it's typed as a list, so they should be in lists!

GOD DAMMIT BARMBULEN

san7890 committed:
ehhh small cleanup may as well

LemonInTheDark commented on code/__DEFINES/traits.dm:
```diff
@@ -114,6 +114,30 @@
 #define HAS_TRAIT_NOT_FROM(target, trait, source) (target.status_traits?[trait] && (length(target.status_traits[trait] - source) > 0))
 
+/// Define for adding multiple traits to a target via a list. Must have a common source and target.
+#define ADD_TRAIT_LIST(target, list_of_traits, source) \
+	do { \
+		if (!islist(list_of_traits)) { \
+			stack_trace("Invalid arguments passed to ADD_TRAIT_LIST! Invoked on [target] with [list_of_traits], source being [source]."); \
+			return; \
+		} \
+		for (var/trait in list_of_traits) { \
+			ADD_TRAIT(target, trait, source); \
+		} \
+	} while (0)
+
+/// Define for removing multiple traits from a target via a list. Must have a common source and target.
+#define REMOVE_TRAIT_LIST(target, list_of_traits, source) \
+	do { \
+		if (!islist(list_of_traits)) { \
+			stack_trace("Invalid arguments passed to REMOVE_TRAIT_LIST! Invoked on [target] with [list_of_traits], source being [source]."); \
+			return; \
+		} \
+		for (var/trait in list_of_traits) { \
+			REMOVE_TRAIT(target, trait, source); \
+		} \
+	} while (0)
```
why are these while wrapped?

san7890 replied:
i did it like that to mirror ADD_TRAIT and REMOVE_TRAIT being while scoped. i couldn't tell you why it's like that on a technical level, but I wanted to ensurethat we didn't really intefere with the operation of the base ADD/REMOVE_TRAIT function

i can run some tests to see if stuff breaks (maybe unit tests) if you think they don't need the while scope

Mothblocks replied:
`ADD_TRAIT` etc are like that because they define a `var/list/_L`, and don't want to bleed it out. You aren't defining any variables so you wouldn't need this

Mothblocks commented on code/__DEFINES/traits.dm:
```diff
@@ -114,6 +114,30 @@
 #define HAS_TRAIT_NOT_FROM(target, trait, source) (target.status_traits?[trait] && (length(target.status_traits[trait] - source) > 0))
 
+/// Define for adding multiple traits to a target via a list. Must have a common source and target.
+#define ADD_TRAIT_LIST(target, list_of_traits, source) \
```
I don't think we are gaining anything from these being macros unless we can prove we get a sizable performance benefit from them, please make them procs in a `__HELPERS/traits.dm` or something (since I think `__DEFINES` should only have defines)

Mothblocks reviewed (changes requested):
Make this a proc and rename it to `add_traits` and `remove_traits`

Mothblocks commented on code/__DEFINES/traits.dm:
```diff
@@ -114,6 +114,30 @@
+/// Define for adding multiple traits to a target via a list. Must have a common source and target.
+#define ADD_TRAIT_LIST(target, list_of_traits, source) \
+	do { \
+		if (!islist(list_of_traits)) { \
+			stack_trace("Invalid arguments passed to ADD_TRAIT_LIST! Invoked on [target] with [list_of_traits], source being [source]."); \
+			return; \
+		} \
```
1. While this is a macro, this definitely should not `return` since that stops the entire proc its in
2. You should use `ASSERT` instead--`ASSERT(islist(list_of_traits), "my message here")`

san7890 committed:
_list -> plural

san7890 committed:
define macro -> proc

san7890 committed:
unit test (oh god oh fuck)

san7890 committed:
cleans up lints

san7890 committed:
removes test_focus

san7890 commented:
i added a unit test mainly because i wanted to make sure mixing/matching ADD_TRAIT/REMOVE_TRAIT with ADD_TRAITS/REMOVE_TRAITS would still work, but i decided I should test all of the "basic" add/remove trait stuff we have juuust in case

Mothblocks commented on code/__HELPERS/traits.dm:
```diff
@@ -28,3 +28,21 @@
 		REMOVE_TRAIT(the_atom2,trait,source)
+
+/// Define wrapper for adding multiple traits to a target via a list. Must have a common source and target.
+#define ADD_TRAITS(target, list_of_traits, source) _add_traits(target, list_of_traits, source)
```
There shouldn't be any macro, there's no point!

Mothblocks replied:
Ideal interface is `add_traits(target, list_of_traits, source)`. There is no need for the indirection

san7890 replied:
keeping it as a macro helps out with linting in case you fumble the arguments somehow (i.e. forgetting a list, implementing multiple sources), which isn't something i believe you can do with procs? i can remove it too but i figured that was neat enough for me to keep the macro

Mothblocks replied:
I don't like that logic because nearly all proc parameters are required

Mothblocks replied:
Ideally we'd have a lint but our tools are not good enough for that (please do not fucking regex this)

san7890 replied:
i also don't like the potential syntax of people trying to `target.add_traits(...)` in the thought that that might do something, very unlikely but it just feels like something that can be confused like that

Mothblocks replied:
I think it *should* be target.add_traits tbh

san7890 committed:
Merge remote-tracking branch 'upstream/master' into definify

san7890 committed:
Merge remote-tracking branch 'upstream/master' into definify

san7890 committed:
makes it a native proc

san7890 committed:
mass regex

san7890 committed:
gets rid of src

san7890 committed:
makes it PascalCase

this is so it fits in better with the other trait macros, as well as match up to AddComponent/AddElement

san7890 committed:
scopes it to datum

global don't work very sad

san7890 committed:
whoopsies i fucked that up

san7890 committed:
fucko bwoingo cleanup

also gets rid of a proc that wasn't necessary anymore

san7890 commented:
i decided it looked better to be PascalCase since it was operating in the same contexts as AddComponent/RemoveComponent etc.

san7890 committed:
Merge remote-tracking branch 'upstream/master' into definify

san7890 committed:
merge conflict fail

Mothblocks reviewed (changes requested):
Last request: We don't do PascalCase anymore, 99% of what is PascalCase now should not be

Mothblocks commented:
(As in, RegisterSignal etc really should be snake_case too)

san7890 committed:
snake case again

san7890 committed:
Merge remote-tracking branch 'upstream/master' into definify

san7890 committed:
cleans up bonus_traits addition

wrapping it in a list makes it so much simpler
